attestation-service/verifier/src/sample/mod.rs

```diff
@@ -0,0 +1,85 @@
+use anyhow::{Context, Result};
+use log::debug;
+extern crate serde;
+use self::serde::{Deserialize, Serialize};
+use super::*;
+use async_trait::async_trait;
+use base64::Engine;
+use serde_json::json;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct SampleTeeEvidence {
+    svn: String,
+
+    #[serde(default = "String::default")]
+    report_data: String,
+
+    #[serde(default = "String::default")]
+    init_data_hash: String,
+}
+
+#[derive(Debug, Default)]
+pub struct Sample {}
+
+#[async_trait]
+impl Verifier for Sample {
+    async fn evaluate(
+        &self,
+        evidence: &[u8],
+        report_data: Option<&[u8]>,
+        init_data_hash: Option<&[u8]>,
+    ) -> Result<TeeEvidenceParsedClaim> {
+        let tee_evidence = serde_json::from_slice::<SampleTeeEvidence>(evidence)
+            .context("Deserialize Quote failed.")?;
+
+        verify_tee_evidence(report_data, init_data_hash, &tee_evidence)
+            .await
+            .context("Evidence's identity verification error.")?;
+
+        debug!("TEE-Evidence<sample>: {:?}", tee_evidence);
+
+        parse_tee_evidence(&tee_evidence)
+    }
+}
+
+async fn verify_tee_evidence(
+    report_data: Option<&[u8]>,
+    init_data_hash: Option<&[u8]>,
+    evidence: &SampleTeeEvidence,
+) -> Result<()> {
+    // Verify the TEE Hardware signature. (Null for sample TEE)
+
+    // Emulate the report data.
+    if let Some(report_data) = report_data {
+        debug!("Check the binding of REPORT_DATA.");
+        let ev_report_data = base64::engine::general_purpose::STANDARD
+            .decode(&evidence.report_data)
+            .context("base64 decode report data for sample evidence")?;
+        if report_data != ev_report_data {
+            bail!("REPORT_DATA is different from that in Sample Quote");
+        }
+    }
+
+    // Emulate the init data hash.
+    if let Some(init_data_hash) = init_data_hash {
+        debug!("Check the binding of init_data_digest.");
+        let ev_init_data_hash = base64::engine::general_purpose::STANDARD
+            .decode(&evidence.init_data_hash)
+            .context("base64 decode init data hash for sample evidence")?;
+        if init_data_hash != ev_init_data_hash {
+            bail!("INIT DATA HASH is different from that in Sample Quote");
+        }
+    }
+
+    Ok(())
```

Comment:
So if report_data and init_data_hash are both `None` the sample verifier is basically a no-op? Can't we require both values to be set like in other verifiers?

Reply:
Yes. This `Sample` verifier is just to provide an example implementation of verifier. Before this PR it only checks the binding of the runtime data. https://github.com/confidential-containers/kbs/blob/main/attestation-service/attestation-service/src/verifier/sample/mod.rs#L46-L58

Why this PR choose to make the parameter `report_data` into `Option<T>` is that the CoCo-AS have enough ability to work as a separate service without KBS to verify the TEE evidences. The service can either help to check the binding of runtime data or not. If yes, the `Option<>` will be `Some`, or it will be `None`.

Reply:
I see, from the perspective of a reference implementation it makes. Since we use the sample verifier in the e2e we should make we pass a nonce in the test, so we test that path.

Reply:
I split the commits. This PR will still ensure the `report_data` be checked when using `sample` verifier.